### ISSUE TITLE
[RFC] Join-based API to support uneven inputs in DDP

### ISSUE THREAD

**Issue by rohan-varma** [pytorch/pytorch#38174]
## 🚀 Feature
with @pritamdamania87 @mrshenli @zhaojuanmao 
This RFC is to summarize the current proposal for supporting uneven inputs across different DDP processes. Related discussion in https://github.com/pytorch/pytorch/issues/33148. An example pain point from a user is on the [PyTorch forums](https://discuss.pytorch.org/t/best-practice-for-uneven-dataset-sizes-with-distributeddataparallel/67308
).
#### Problem
[torch.nn.parallel.DistributedDataParallel](https://pytorch.org/docs/stable/nn.html#torch.nn.parallel.DistributedDataParallel) is a commonly used tool for distributed data-parallel training, but currently obliges the user to provide an equal number of inputs across each participating DDP process (or appropriately handle the error otherwise). DDP currently fails when different processes have an unequal number of inputs to process during training. While there are utilities such as [DataLoader](https://pytorch.org/docs/stable/data.html) and [DistributedSampler](https://pytorch.org/docs/stable/data.html#torch.utils.data.distributed.DistributedSampler) that make navigating this assumption in DDP easier by evenly distributing the dataset, we can't expect these to solve all use cases and many users have had use cases where uneven inputs need to be supported.

The following script gives a simple example of the error:

```
import torch
import torch.distributed as dist
import os
import torch.multiprocessing as mp
import torch.nn as nn
def worker(rank):
    dist.init_process_group("nccl", rank=rank, world_size=2)
    torch.cuda.set_device(rank)
    model = nn.Linear(1, 1, bias=False).to(rank)
    model = torch.nn.parallel.DistributedDataParallel(model, device_ids=[rank], output_device=rank)
    # Create uneven inputs, rank 1 will get one more input than rank 0. This will cause a hang.
    inputs = [torch.tensor([1]).float() for _ in range(10 + rank)]
    for _ in range(5):
        for inp in inputs:
            loss = model(inp).sum()
            loss.backward()
    torch.cuda.synchronize(device=rank)

if __name__ == '__main__':
    os.environ["MASTER_ADDR"] = "localhost" ; os.environ["MASTER_PORT"] = "29501"
    mp.spawn(worker, nprocs=2, args=())
```

With the NCCL backend (recommended choice when training with multiple GPUs) this will result in a hang, as process 1 will wait for communication (allreduce) from process 0, but process 0 has already exited its training loop. On the other hand, with the gloo backend, this results in a "Connection reset by peer" error.

#### Proposal

This was proposed by @pritamdamania87 and is influenced by the approach taken by Horovod to resolve a similar problem (https://github.com/horovod/horovod/issues/832).
1. Provide a context manager such as `with torch.nn.parallel.distributed.join()`. In the `__enter__`, we will set a flag indicating that we will run the below process for managing uneven inputs.
2. The context manager's `__exit__` indicates that the process has depleted its input and is ready to join. When a trainer calls this:
         a. Schedule an allreduce with `torch.tensor(0)`. This allreduce will match the allreduce scheduled by non-joined processes (explained below in point 3)
         b. If the result of the above is zero, this means that all processes have depleted their inputs and we can move to step (d)
         c. Otherwise, schedule an allreduce for all buckets in the backwards pass, with all gradients zeroed out (this is so that joined ranks don't affect gradients of the rest of the training processes). This will match the allreduce done in the backwards pass for currently active trainers. Go back to step a.
    d. If (a) returns all zeros, this means that all ranks have terminated their inputs and we can move on to cleanup. We also need to keep track of the process with the latest model parameters, and broadcast them to all ranks to maintain the fact that DDP ensures all parameters across processes are the same. We can do this via a simple version counter. In this step, we can then allgather this version counter, and have the process with the maximum counter broadcast its parameters to the rest of the processes. Ties can be broken arbitrarily.

3. If a trainer has not called `__exit__`, then:
    a. Before scheduling allreduces for the backwards pass, schedule an allreduce with `torch.tensor(1)`. This allreduce matches the one scheduled in (2a). We can schedule this allreduce in the forward pass, but we should not await it here for performance reasons; it should be awaited at the end of the backwards pass.
    b. Schedule allreduce ops for all the buckets as typical in the backwards pass for DDP. Processes which have depleted their inputs will match these allreduces as a result of step 2c.  These processes will have zero as the argument for their gradients so they will not contribute to gradient averaging.
    c. Instead of dividing by a static world_size, since we now can have a smaller effective world size (initial_world_size - currently_joined_processes), divide by this instead to ensure that we are still correctly averaging gradients. This can be done by taking the value returned in 3a, which will be interpreted as an int representing the number of currently active processes. 

#### Code sample

```
model = nn.Linear(1, 1, bias=False).to(rank)
    model = torch.nn.parallel.DistributedDataParallel(model, device_ids=[rank], output_device=rank)
    # Create uneven inputs
    inputs = [torch.tensor([1]).float() for _ in range(10 + rank)]
    for _ in range(epochs):
        with torch.nn.parallel.distributed.join():
            for inp in inputs:
                loss = model(inp).sum()
                loss.backward()
```

#### Alternatives considered
We considered the alternative of all trainers raising a `StopIteration` once we detect that at least one trainer has depleted its input (via the above method). However, the user would then have to catch this StopIteration and this would also result in all processes stopping their training, whereas the currently proposed method allows training to continue with a smaller effective world size. In the future if we see the need for users to actually stop the training early in these situations, we can provide the appropriate options, although we would like to keep usage of this API as simple as reasonably possible.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @xush6528 @osalpekar

**Comment by pritamdamania87** [comment]
A slight improvement here would be to avoid waiting for 2.b and instead schedule both 2. b and 2. c together and wait for all results together. This would ensure ranks that are not in `join` are not waiting too long for the allreduce in 2. c. Also, this is symmetric to what we do for ranks not in ddp.join(): They schedule the torch.tensor(1) op and the bucket allreduce ops together and wait for all of them together.

**Comment by ppwwyyxx** [comment]
Note that certain models (e.g., those with nn.SyncBatchNorm, or the recent MoCo paper from FAIR) have collective calls during forward as well. Those collective calls in forward will also hang with uneven inputs.
It doesn't look like the current proposal is taking these models into account.

Another comment: a simple alternative that's perhaps more robust to any such hangs, is to raise a `TimeoutException` after user-specified timeout period, and let the user handle it. I'm not sure if this is already available now on all backends, but it seems worth discussing as an alternative.

**Comment by soulslicer** [comment]
I wanted to post this thread to show how it is solved:

https://discuss.pytorch.org/t/multiprocessing-barrier-blocks-all-processes/80345/29

**Comment by pritamdamania87** [comment]
@ppwwyyxx This is a good point that we overlooked during the initial design and you're right that this won't work if there are other collective operations in the forward pass. We do have a timeout for collective operations and GLOO raises an appropriate exception, although for NCCL this is only enabled when `NCCL_BLOCKING_WAIT` is set.

**Comment by pritamdamania87** [comment]
An alternative I was thinking about was as follows:

1. We have the allreduce operations mentioned in 2.a and 3.a in the proposal and run them at the very beginning of the forward pass.
2. We wait for these allreduce operations after the backward pass is done.
3. If the result is not equal to the world size, it means some trainers have exited. In this case, the other trainers abort all pending collective operations and throw `StopIteration`.

Although, one problem with this approach is that DDP might not be aware of the process_group that is used for collective operations of SyncBatchNorm. As a result, we can't abort those collective operations.

@zhaojuanmao @rohan-varma @mrshenli Was wondering if you had any other ideas? It does seem like having the user deal with this might be a simple alternative.

**Comment by rohan-varma** [comment]
@pritamdamania87 Wouldn't that proposal also result in us being unable to continue training, while the described proposal here makes that possible? Maybe we could have some sort of a hybrid solution where collective halls in the forward pass raise a timeout that the user has to manage instead of hanging in the forward pass. Or, we can support handling collective hangs in the forward pass in a similar manner that we do for backwards, and we'd have to consider how to make DDP aware of those forward calls and the underlying PG. 

I think that supporting scenarios in which the user does not want to abort training and continue with the remaining workers is still useful and provides a more flexible API.

**Comment by pritamdamania87** [comment]
> Wouldn't that proposal also result in us being unable to continue training, while the described proposal here makes that possible?

Yes this is correct, although the problem with the current proposal is we can't deal with cases where the forward pass has some collective calls. Even the proposal that I mentioned above has an issue that we can't reliably identify all the collective calls to abort.

>  Or, we can support handling collective hangs in the forward pass in a similar manner that we do for backwards, and we'd have to consider how to make DDP aware of those forward calls and the underlying PG.

There are a couple of hard to resolve issues here:

1) Its hard for DDP to be aware of collective calls in the forward pass. We could do some processing to detect things like SyncBatchNorm in the forward pass, but there could be other custom user modules that perform collective communication and we wouldn't be aware of those.
2) Let's say we somehow are able to detect all the collective calls, the problem now is how do trainers who have completed training know what data to send in their "dummy" collective calls? In case of allreduce we had special handling to send zeros and divide only by the currently active trainers. But for things like SyncBatchNorm and other generic collectives that the user might use, it's not clear what might be the right data to send over.

> Maybe we could have some sort of a hybrid solution where collective halls in the forward pass raise a timeout that the user has to manage instead of hanging in the forward pass

I agree with this, I feel like vast majority of cases wouldn't have collective calls in the forward pass and this might still be useful for those cases.

**Comment by soulslicer** [comment]
https://discuss.pytorch.org/t/multiprocessing-barrier-blocks-all-processes/80345/31

This person helped me solve it

**Comment by zhaojuanmao** [comment]
> An alternative I was thinking about was as follows:
> 
> 1. We have the allreduce operations mentioned in 2.a and 3.a in the proposal and run them at the very beginning of the forward pass.
> 2. We wait for these allreduce operations after the backward pass is done.
> 3. If the result is not equal to the world size, it means some trainers have exited. In this case, the other trainers abort all pending collective operations and throw `StopIteration`.
> 
> Although, one problem with this approach is that DDP might not be aware of the process_group that is used for collective operations of SyncBatchNorm. As a result, we can't abort those collective operations.
> 
> @zhaojuanmao @rohan-varma @mrshenli Was wondering if you had any other ideas? It does seem like having the user deal with this might be a simple alternative.

@pritamdamania87 if forward pass hangs, we can not go to step2 here to wait for all reduce, right?

**Comment by zhaojuanmao** [comment]
If users would like to handle uneven inputs, they can try different ways, as @mrshenli  mentioned in the forum: 1) if know the number of inputs, they will know there is uneven inputs or not, and drop the tail data  2) if not know the number of input, they can pipeline data reading and identify end of data on their own, then drop the tail data

For DDP API, I think this proposal provides another option (continue training for uneven inputs) for users and will be useful for quite a lot of cases, so voting for keep current design in this proposal.

**Comment by yaadhavraajagility** [comment]
Is there any example on how to use this?

**Comment by chenjiasheng** [comment]
Can you please make this also work for occaisional OOM?

**Comment by Huiimin5** [comment]
Could you please share how you solved this problem?

> Can you please make this also work for occaisional OOM? E.g. drop the bad batch of the bad rank.

**Comment by rohan-varma** [comment]
@yaadhavraajagility we have a tutorial: https://pytorch.org/tutorials/advanced/generic_join.html#:%7E:text=The%20context%20manager%20allows%20the,shadowed%20are%20specified%20by%20hooks

@chenjiasheng: Great suggestion, please file an issue describing your use case and details about what you'd like to see in the feature!
